### PR TITLE
Fix/resolve atf add sub menu known issues

### PR DESCRIPTION
**Pull request by BrandonHe**
Please help review @AByzhynar @pestOO @justinjdickow @anosach-luxoft 
Replace PR #654, change branch name from hotfix to fix

### Code Update:
- RequestController::UpdateTimer():<br>
  Change the timer update logic from waiting until close and restart to unblock the waiting threads, avoid timeout issue
- RequestController::OnTimer():<br>
  Add Autolock and wait to process the condition of NOT probably_expired
- SmartObject::cleanup_data():<br>
  Set Smart Object clean up data to NULL after it's deleted

### Verify Test:
- PASSED: Both ATF_AddSubMenu and ATF_SetMediaClockTimer ATF tests, do the test several times
- PASSED: SPT AddSubMenu function test(successfully add ten submenu)
- PASSED: Unit test

**Attached test result:**
[unit_test_reuslt.txt](https://github.com/smartdevicelink/sdl_core/files/342775/unit_test_reuslt.txt)
[ATF_AddSubMenu_fix_test_result.txt](https://github.com/smartdevicelink/sdl_core/files/342777/ATF_AddSubMenu_fix_test_result.txt)
[ATF_SetMediaClockTimer_fix_test_result.txt](https://github.com/smartdevicelink/sdl_core/files/342778/ATF_SetMediaClockTimer_fix_test_result.txt)

### CODE REVIEW

File: src/components/smart_objects/src/smart_object.cc [hunk 1]
```diff
@@ -723,16 +723,28 @@ void SmartObject::duplicate(const SmartObject& OtherObject) {
 void SmartObject::cleanup_data() {
   switch (m_type) {
     case SmartType_String:
-      delete m_data.str_value;
+      if (m_data.str_value) {
```
Comment (AGaliuzov) [inline src/components/smart_objects/src/smart_object.cc]: @BrandonHe What the purpose of this `if` condirion and `NULL` assignment?

**Comment by AGaliuzov** [comment]
@LuxoftAKutsan please participate

File: src/components/application_manager/src/request_controller.cc [hunk 2]
```diff
@@ -364,9 +369,27 @@ void RequestController::onTimer() {
   SDL_AUTO_TRACE();
   SDL_DEBUG(
       "ENTER Waiting fore response count: " << waiting_for_response_.Size());
-  RequestInfoPtr probably_expired =
-      waiting_for_response_.FrontWithNotNullTimeout();
-  while (probably_expired && probably_expired->isExpired()) {
+  while(!stop_flag_) {
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/request_controller.cc]: @BrandonHe  in this realisation timer will work just like a thread. It is confuting to use timer for thread like functionality. 
If you think that threadlike functionality will be better, please rewrite request controller, remove timer and use thread. But I don't think that it is a good Idea.

**Comment by LuxoftAKutsan** [comment]
@BrandonHe please write root cause of defects. Because this changes rather confusing.

File: src/components/application_manager/src/request_controller.cc [hunk 3]
```diff
@@ -364,9 +369,27 @@ void RequestController::onTimer() {
...
+  while(!timer_stop_flag_) {
+    RequestInfoPtr probably_expired =
+        waiting_for_response_.FrontWithNotNullTimeout();
+    if (!probably_expired) {
+      sync_primitives::AutoLock auto_lock(timer_lock);
+      timer_condition_.Wait(auto_lock);
+      continue;
+    }
+    if (!probably_expired->isExpired()) {
+      SDL_INFO("onTimer:isExpired");
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/request_controller.cc]: @BrandonHe  Please add additional information about request that expired

File: src/components/application_manager/src/request_controller.cc [hunk 4]
```diff
@@ -364,9 +369,27 @@ void RequestController::onTimer() {
...
+      sync_primitives::AutoLock auto_lock(timer_lock);
+      const TimevalStruct current_time = date_time::DateTime::getCurrentTime();
+      const TimevalStruct end_time = probably_expired->end_time();
+      if (current_time < end_time) {
+        const uint32_t msecs = static_cast<uint32_t>(
+            date_time::DateTime::getmSecs(end_time - current_time));
+        SDL_DEBUG("Sleep for" << msecs << " millisecs");
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/request_controller.cc]: @BrandonHe space after for

File: src/components/smart_objects/src/smart_object.cc [hunk 5]
```diff
@@ -723,16 +723,28 @@ void SmartObject::duplicate(const SmartObject& OtherObject) {
 void SmartObject::cleanup_data() {
```
Comment (LuxoftAKutsan) [inline src/components/smart_objects/src/smart_object.cc]: @BrandonHe It is very strange to see modification of logic request controller and smart object in one pull request. Please describe what for do you need this changes.

File: src/components/smart_objects/src/smart_object.cc (hunk 1 above)
Comment (BrandonHe) [inline src/components/smart_objects/src/smart_object.cc]: @AGaliuzov You know, in C plus, if value in ```if``` condition is ```NULL```, operator ```delete``` should do nothing, check the value that to delete is a good style in C plus. And Also, there is a very small possibility that delete fails, that may cause memory leak. Third, this is to avoid of dangling pointer.

File: src/components/application_manager/src/request_controller.cc (hunk 2 above)
Comment (BrandonHe) [inline src/components/application_manager/src/request_controller.cc]: @LuxoftAKutsan This is not rewrite, actually the timer logic is not changed here,  here add ```AutoLock``` to make it works better in multi-thread program. And during the test, I usually found ```TIME_OUT`` error based on the machine performance.

File: src/components/smart_objects/src/smart_object.cc (hunk 5 above)
Comment (BrandonHe) [inline src/components/smart_objects/src/smart_object.cc]: @LuxoftAKutsan Yeah, this modification is not relative with logic request controller, because this modification is too small changes to send another PR, so I sent this PR here, please reference [comment](https://github.com/smartdevicelink/sdl_core/pull/660#discussion_r69402044). If really need send another PR for this, let me know. Thanks!

File: src/components/application_manager/src/request_controller.cc (hunk 4 above)
Comment (BrandonHe) [inline src/components/application_manager/src/request_controller.cc]: done with new commit [9a60a08](https://github.com/BrandonHe/sdl_core/commit/9a60a08a6037fb6c47b1aabb885fa6adf8e9516a)

File: src/components/application_manager/src/request_controller.cc (hunk 3 above)
Comment (BrandonHe) [inline src/components/application_manager/src/request_controller.cc]: done with new commit [9a60a08](https://github.com/BrandonHe/sdl_core/commit/9a60a08a6037fb6c47b1aabb885fa6adf8e9516a)

**Comment by LuxoftAKutsan** [comment]
@BrandonHe  please note that check style fails. 
And describe  root cause of defect.

File: src/components/application_manager/src/request_controller.cc [hunk 6]
```diff
@@ -364,9 +369,27 @@ void RequestController::onTimer() {
   SDL_AUTO_TRACE();
   SDL_DEBUG(
       "ENTER Waiting fore response count: " << waiting_for_response_.Size());
-  RequestInfoPtr probably_expired =
-      waiting_for_response_.FrontWithNotNullTimeout();
-  while (probably_expired && probably_expired->isExpired()) {
+  while(!timer_stop_flag_) {
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/request_controller.cc]: @BrandonHe I can see that onTimer function will be called once, and then it will be just looped in condition variable while it won't be stopped in destructor. I can't understand why do you use timer for this purpose, you should use thread logic for this purpose. You just implemented timer inside timer callback, it is confusing. 

Also within you realisation "expired logic" will be executed each time new request is added (because of NotifyOne in UpdateTimeout), it is redundant.

File: src/components/smart_objects/src/smart_object.cc (hunk 1 above)
Comment (AGaliuzov) [inline src/components/smart_objects/src/smart_object.cc]: @BrandonHe 
1.  C++ standard allows NULL-pointer deletion.
Please see C++98 standard 5.3.5 Delete paragraph 2.
> In either alternative, if the value of the operand of `delete` is null pointer the operation has no effect

2.
> And Also, there is a very small possibility that delete fails, that may cause memory leak

Could you please let me know in which way `if-condition` helps to prevent memory leak?

>Third, this is to avoid of dangling pointer.

To avoid dangled pointer you are using `NULL` assignment which is good.

So, all I want is to avoid redundant `if-condition`

File: src/components/application_manager/src/request_controller.cc (hunk 6 above)
Comment (BrandonHe) [inline src/components/application_manager/src/request_controller.cc]: @LuxoftAKutsan The while condition can also be stopped by ```break``` besides in destructor, 
This purpose is to keep sure whenever this onTimer is called,  the current thread both under condition ```!expired``` and under condition ```!probably_expired->isExpired()``` can be blocked to wake up(previous lack of this purpose). Other executed logic is same with original condition ```(probably_expired && probably_expired->isExpired())```. I am not very clearly understand your thread logic

File: src/components/smart_objects/src/smart_object.cc (hunk 1 above)
Comment (BrandonHe) [inline src/components/smart_objects/src/smart_object.cc]: @AGaliuzov 
> 1. C++ standard allows NULL-pointer deletion.
> Please see C++98 standard 5.3.5 Delete paragraph 2.

I didn't mean that delete NULL-pointer is not allowed, I just mean if is already NULL, the program will no nothing, in case delete two times which is a disaster.
> 2. Could you please let me know in which way if-condition helps to prevent memory leak?

I mean this not a reason for ```if``` but for ```NULL```
Thanks!

Comment (AGaliuzov) [inline src/components/smart_objects/src/smart_object.cc]: @BrandonHe Al I try to say is that `if-condition` has no ane sense here. Please remove it

Comment (BrandonHe) [inline src/components/smart_objects/src/smart_object.cc]: @AGaliuzov 
Done in commit [`461e69f`](https://github.com/smartdevicelink/sdl_core/pull/660/commits/461e69fc7510ed435e71a6aa0e00242a128221ff)

**Comment by LuxoftAKutsan** [comment]
@BrandonHe I still cant figure out root cause of problem that you tried to solve

**Comment by BrandonHe** [comment]
@LuxoftAKutsan 
OK, Here is more detailed infos, hopes let you know the cause.
I tried to solve in this PR is that during the ATF_AddSubMenu testing, we found test fails as well as in [KnownIssues.md](https://github.com/smartdevicelink/sdl_atf_test_scripts/blob/master/KnownIssues.md)

We analysed and found it caused by timeout. As you know, ```RequestController``` is responsible for handling timeouts of responsibles and notifications from HMI. 

The previous ```UpdateTimer``` is just wait a fixed time to close the Timer and restart a new one, BUT program here relate to waiting and waking up of multi-threading, the actual wait time may longer than that fixed time. It will cause timeout.

**Comment by BrandonHe** [comment]
@AGaliuzov @LuxoftAKutsan 
Do you have any other feedback?

**Comment by LuxoftAKutsan** [comment]
@BrandonHe But how your fix remove issue that next real timeout may  y be called a little bit later then fixed one?

**Comment by BrandonHe** [comment]
@LuxoftAKutsan 
By adding auto_lock here:
```
+      sync_primitives::AutoLock auto_lock(timer_lock);
+      timer_condition_.Wait(auto_lock)
...
...
+      timer_condition_.NotifyOne();
```
BTW, we run all ATF test scripts on limited performance port(such as WinCE), most failed results are timeout.

File: src/components/application_manager/src/request_controller.cc (hunk 2 above)
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/request_controller.cc]: @BrandonHe I am trying to understand new flow. I can see that you mixed timers and loops. There is simplified current logic of RequestCtrl [sequence](https://www.websequencediagrams.com/cgi-bin/cdraw?lz=dGl0bGUgUmVxdWVzdCBjb250cm9sbGVyCgoKQU0tPgASB0N0cmw6IEFkZAAiCAoADgsAFw9QdXQgcgBJB2luIHF1ZXVlIAAiDlRpbWVyOiBVcGRhdGUgdGltZXIKAA8FAAUWAA18RG8gbm90IHUAgRgTAIIPDU9uVGltZW91AIF9HXJvY2VzcyBmaXJzAIISEgCBZj0APVYgKGxhc3Qgb25lKQCDLBVTdXNwZW5kAIMtFQAODgo&s=earth) Could you create sequence diagram with new logic?

Comment (LuxoftAKutsan) [inline src/components/application_manager/src/request_controller.cc]: @BrandonHe  do you have any updates?

Comment (BrandonHe) [inline src/components/application_manager/src/request_controller.cc]: @LuxoftAKutsan Sorry for delay reply, I missed the mail. New sequence diagram please reference [**here**](https://www.websequencediagrams.com/cgi-bin/cdraw?lz=dGl0bGUgUmVxdWVzdCBDb250cm9sbGVyCgpBTS0-ABEHQ3RybDogQWRkACEICgAOCwAXD3B1dCByAEgHaW4gcXVldWUAIQ5UaW1lcjogU3RhcnQgAAgFCgpsb29wIG5vdCB0aW1lcl9zdG9wX2ZsYWcKCm9wdAA_BiBpcyBlbXB0eQoAPAUAPwlBdXRvIGxvY2sgd2FpdAplbmQALwZub3QgaXNFeHBpcmVkACUPd2FpdCBmb3IgbXNlY3MALwUAMwUAgRQUVXBkYXRlAIETBgBsD1dhaXQAgXE4AIF_KQBgIACCBwcAgx4Nb25UaW1lb3V0AIFEDwCDQw1Qcm9jZXNzIGZpcnMAAYELAHkTIChsYXN0IG9uZSkAhDUWdXNwZW5kAIMPFQCEEA8AhSAaRXhpAIVJDwCFIgcAhQQPID0gdHJ1ZQoK&s=rose), I think the most logic is not changed.

Comment (BrandonHe) [inline src/components/application_manager/src/request_controller.cc]: @LuxoftAKutsan Any other concerns do you have?

**Comment by LuxoftAKutsan** [comment]
@BrandonHe could you create Unit Test where you can reproduce defect in request controller that will be passed after your changes?

**Comment by BrandonHe** [comment]
@LuxoftAKutsan Sorry, I am not familiar with create Unit Test:disappointed:. But I think it's unnecessary to rewrite unit test code for this PR as most logic is not changed, supposed you have compared [previous sequence](https://www.websequencediagrams.com/cgi-bin/cdraw?lz=dGl0bGUgUmVxdWVzdCBjb250cm9sbGVyCgoKQU0tPgASB0N0cmw6IEFkZAAiCAoADgsAFw9QdXQgcgBJB2luIHF1ZXVlIAAiDlRpbWVyOiBVcGRhdGUgdGltZXIKAA8FAAUWAA18RG8gbm90IHUAgRgTAIIPDU9uVGltZW91AIF9HXJvY2VzcyBmaXJzAIISEgCBZj0APVYgKGxhc3Qgb25lKQCDLBVTdXNwZW5kAIMtFQAODgo&s=earth) and [updated sequence](https://www.websequencediagrams.com/cgi-bin/cdraw?lz=dGl0bGUgUmVxdWVzdCBDb250cm9sbGVyCgpBTS0-ABEHQ3RybDogQWRkACEICgAOCwAXD3B1dCByAEgHaW4gcXVldWUAIQ5UaW1lcjogU3RhcnQgAAgFCgpsb29wIG5vdCB0aW1lcl9zdG9wX2ZsYWcKCm9wdAA_BiBpcyBlbXB0eQoAPAUAPwlBdXRvIGxvY2sgd2FpdAplbmQALwZub3QgaXNFeHBpcmVkACUPd2FpdCBmb3IgbXNlY3MALwUAMwUAgRQUVXBkYXRlAIETBgBsD1dhaXQAgXE4AIF_KQBgIACCBwcAgx4Nb25UaW1lb3V0AIFEDwCDQw1Qcm9jZXNzIGZpcnMAAYELAHkTIChsYXN0IG9uZSkAhDUWdXNwZW5kAIMPFQCEEA8AhSAaRXhpAIVJDwCFIgcAhQQPID0gdHJ1ZQoK&s=rose). Also, I have run unit test after changing the codes, and all passed, you can see result in **Attached test result:** in my first comment.

**Comment by LuxoftAKutsan** [comment]
@BrandonHe As I understand there were some problem in Request controller, so you made this changes. We need unit test to be sure that this problem won't repeated in future at least.  And Unit test would be rather good approve that this problem existed without this changes and not it is not reproduces. ATF is very  common check, area of defects found by ATF is rely big. if any code is possible to check with UT, we should check it with UT. 
And any way UT is part of pull request [checkout ](https://github.com/smartdevicelink/sdl_core/blob/develop/COMMITTERS.md#pull-request-checklisthttps://github.com/smartdevicelink/sdl_core/blob/develop/COMMITTERS.md#pull-request-checklist )
There are already  have some unit tests for request controller, you can see how did they done and create new one.

**Comment by BrandonHe** [comment]
@LuxoftAKutsan After research, I think the present UT has covered this changed in [request_controller_test.cc](https://github.com/smartdevicelink/sdl_core/blob/develop/src/components/application_manager/test/request_controller/request_controller_test.cc). This is no parameters or return values changed in ```void RequestController::onTimer()``` and in```void RequestController::UpdateTimer()```, hence, I think there is no need to add new extra Mock or Expect behaviors. Could you give more instructions if you still think need this?

**Comment by LuxoftAKutsan** [comment]
@BrandonHe I wan't to find out what exactly problem did you fix. So I would like you to write Unit test that will fail with old realisation of RequestCtrl and will not fail on current.

**Comment by LuxoftAKutsan** [comment]
@BrandonHe I am reading this changes. 
I can see, that in case if probably expired request in not expired, you prefer to wait on condition variable while it won't be expired. UpdateTimer function and timer implementation will actually do this. 
UpdateTimer will take FronWithNotNullTimeout and start timer for this value.  Timer will wait this time on condition variable and call onTimer again. 
So I can't figure out what problem does it fix. You just copied part of timer realisation inside onTimer method. Not it looks very very complicated because of couple of continue and breaks inside cycle, big amount of code in onTimer method, mixing of condition variables usages and Timer to find expired request.

**Comment by LuxoftAKutsan** [comment]
@BrandonHe  Do you have any updates?

**Comment by BrandonHe** [comment]
@LuxoftAKutsan 
Sorry for ignore your comment as received too much mail from smartdevicelink repo every day. Yeah, you are exactly right about what i did in this PR. The original reason is that I tried to fix ATF_AddSubMenu fails in [KnowIssues.md](https://github.com/smartdevicelink/sdl_atf_test_scripts/blob/master/KnownIssues.md), I analysed that may cause by timeout, so I updated the codes and **PASSED** ATF_AddSubMenu test as you see in the PR description attachment [ATF_AddSubMenu_fix_test_result.txt](https://github.com/smartdevicelink/sdl_core/files/342777/ATF_AddSubMenu_fix_test_result.txt)

But after reading code again and again, I think you are right, it seems has same logic with previous code, and may make code more complicated. Anyway. I think there is also no new issue added.

**Comment by AGaliuzov** [comment]
@BrandonHe In this case do we still need this pull request if it
>seems has same logic with previous code, and may make code more complicated.

**Comment by BrandonHe** [comment]
@AGaliuzov @LuxoftAKutsan 
Could you please have a discussion, and let me know if i need to close this PR. Thanks!

**Comment by LuxoftAKutsan** [comment]
@BrandonHe if there is not changes in logic and code became more complicated, you can close  this pull request. 

I suppose that ATF_AddSubMenu test fails due to some other reason, and you changes timings that probably shadows fails of ATF_AddSubMenu test. This is just my subjection, and fails of ATF_AddSubMenu test  need additional investigation.

**Comment by BrandonHe** [comment]
OKay.
Close this pull request, need further investigation

**Comment by BrandonHe** [comment]
@AGaliuzov @LuxoftAKutsan @JackLivio @madhuy 
Reopen this PR since this fix is effective to resolve issue #479 , please see comment in #479.
Could you guys review again?

**Comment by LuxoftAKutsan** [comment]
@BrandonHe we have same situation as before. Can you find out root case? Why with previous implementation of RequestCtrl ATF tests failed, but with new does not. I can see that you copied part of  timer implementation(loop condition variable and stop flag staff) in RequestCtrl OnTimer method and I can't see how it can help because it doesn't change logic, just made it more complex.

**Comment by jacobkeeler** [comment]
@BrandonHe Small note: if this fixes issues that currently exist in master, as seems to be stated by @madhuy in #479 and #621, then this should be set up as a hotfix and be based off of master.

I actually have already performed the rebasing as part of an attempt to test your changes, so I could push the rebased version to your fork under a new branch if that is alright.

**Comment by BrandonHe** [comment]
@LuxoftAKutsan As mentioned in previous comments thread, I still think the root cause is not use ```auto_lock``` method in process of each request. It may time out if there have many requests to respond but SDL can not process them as quick as within timeout value. So I use loop to keep sure each request executes  ```timer_condition_.WaitFor(auto_lock, msecs)```. Alternatively, we can change the configuration in smartdevicelink.ini to set a longer time out value, but it's a workaround. Both Android and iOS has this problem. All in all, the most logic in request controller is not changed here I think.

Attached sequence diagram again for all of you reference:
**previous sequence:**
request controller sequence diagram

-------
**update the sequence:**
request controller sequence diagram updated by brandon

**Comment by BrandonHe** [comment]
@jacobkeeler OK, thanks. I am not very familiar with rule, so I changed branch name from hotfix to fix. anything here need I to do, let me know.

**Comment by jacobkeeler** [comment]
@BrandonHe I don't see any changes in your SDL fork. 

This branch is still based off of develop, and it needs to be based off of master since it fixes existing issues in the current release of SDL. Since this will likely need to be done on a new branch, a new PR will likely need to be created. 

I have already performed the rebase of these changes and created [this new branch](https://github.com/smartdevicelink/sdl_core/tree/hotfix/resolve_ATF_AddSubMenu_KnownIssues) on the main repo.

**Comment by BrandonHe** [comment]
@jacobkeeler OK, I will branch off from master to create a new hotfix branch, rebase these changes and send a new pull request later.

**Comment by jacobkeeler** [comment]
@BrandonHe By the way, I attempted to run the tests that this PR was originally made to fix (ATF_AddSubMenu.lua), and none of the AddSubMenu tests mentioned in [KnownIssues.md](https://github.com/smartdevicelink/sdl_atf_test_scripts/blob/master/KnownIssues.md) fail on the master branch as it stands.

Do these tests only fail under special conditions, or were they already fixed in master through another PR?

From what I can understand, this PR is meant to take care of cases where requests can timeout despite getting a response in time, because they are unable to process them before time runs out. Is this correct?

**Comment by madhuy** [comment]
@jacobkeeler @BrandonHe  Please prioritize this PR.

**Comment by BrandonHe** [comment]
@madhuy Resend PR #909 
@jacobkeeler For preconditions, I noticed you had discussion in issue #621, maybe you can also reference #479, I run across issue very similar with #479. I did not send another PR.
>
>From what I can understand, this PR is meant to take care of cases where requests can timeout despite getting a response in time, because they are unable to process them before time runs out. Is this correct?

Yes, this is correct. I originally try to resolve "GENERIC_ERROR" caused by ATF test "ATF_AddSubMenu.lua", especially plenty AddSubMenu requests in very short time alike stress testing.

**Comment by jacobkeeler** [comment]
Closed in favor of #909
